File: laba5/ex1.cpp
#include <iostream>
#include <map>
#include <vector>
#include <string>
#include <sstream>
#include <iomanip>

using namespace std;

struct Cell {
    string item;
    int quantity;
};

map<string, Cell> warehouse;

void initializeWarehouse() {
    vector<char> zones = {'A', 'B', 'C'}; // Исправлены кириллические символы
    for (char zone : zones) {
        for (int shelf = 1; shelf <= 20; ++shelf) {
            for (int section = 1; section <= 5; ++section) {
                for (int rack = 1; rack <= 2; ++rack) {
                    // Формирование адреса без ведущего нуля для стеллажей
                    string address = string(1, zone) 
                                   + to_string(shelf) 
                                   + to_string(section) 
                                   + to_string(rack);
                    warehouse[address] = {"", 0};
                }
            }
        }
    }
}

void processAddCommand(const vector<string>& tokens) {
    if (tokens.size() != 4) {
        cout << "Ошибка: неверный формат команды. Используйте: ADD <товар> <количество> <адрес>" << endl;
        return;
    }
    
    string item = tokens[1];
    int quantity;
    try {
        quantity = stoi(tokens[2]);
        if (quantity <= 0) throw invalid_argument("Ошибка");
    } catch (...) {
        cout << "Ошибка: количество должно быть положительным числом" << endl;
        return;
    }
    
    string address = tokens[3];
    
    if (warehouse.find(address) == warehouse.end()) {
        cout << "Ошибка: ячейка " << address << " не существует" << endl;
        return;
    }

    Cell& cell = warehouse[address];
    if (cell.quantity == 0) {
        if (quantity > 10) {
            cout << "Ошибка: в ячейку нельзя добавить более 10 единиц" << endl;
            return;
        }
        cell.item = item;
        cell.quantity = quantity;
        cout << "Добавлено " << quantity << " ед. товара \"" << item << "\" в " << address << endl;
    } else {
        if (cell.item != item) {
            cout << "Ошибка: ячейка содержит другой товар (\"" << cell.item << "\")" << endl;
            return;
        }
        if (cell.quantity + quantity > 10) {
            cout << "Ошибка: превышен лимит ячейки. Доступно место: " << 10 - cell.quantity << endl;
            return;
        }
        cell.quantity += quantity;
        cout << "Обновлено: " << item << " x" << cell.quantity << " в " << address << endl;
    }
}

void processRemoveCommand(const vector<string>& tokens) {
    if (tokens.size() != 4) {
        cout << "Ошибка: неверный формат команды. Используйте: REMOVE <товар> <количество> <адрес>" << endl;
        return;
    }
    
    string item = tokens[1];
    int quantity;
    try {
        quantity = stoi(tokens[2]);
        if (quantity <= 0) throw invalid_argument("Ошибка");
    } catch (...) {
        cout << "Ошибка: количество должно быть положительным числом" << endl;
        return;
    }
    
    string address = tokens[3];
    
    if (warehouse.find(address) == warehouse.end()) {
        cout << "Ошибка: ячейка " << address << " не существует" << endl;
        return;
    }

    Cell& cell = warehouse[address];
    if (cell.quantity == 0) {
        cout << "Ошибка: ячейка " << address << " пуста" << endl;
        return;
    }
    if (cell.item != item) {
        cout << "Ошибка: в ячейке находится другой товар (\"" << cell.item << "\")" << endl;
        return;
    }
    if (cell.quantity < quantity) {
        cout << "Ошибка: недостаточно товара. Доступно: " << cell.quantity << endl;
        return;
    }
    
    cell.quantity -= quantity;
    if (cell.quantity == 0) cell.item = "";
    cout << "Удалено " << quantity << " ед. товара \"" << item << "\" из " << address 
         << ". Остаток: " << cell.quantity << endl;
}

void processInfoCommand() {
    int totalItems = 0;
    map<char, int> zoneItems;
    vector<string> occupiedCells, emptyCells;

    for (const auto& [address, cell] : warehouse) {
        if (cell.quantity > 0) {
            totalItems += cell.quantity;
            zoneItems[address[0]] += cell.quantity;
            occupiedCells.push_back(address + ": \"" + cell.item + "\" x" + to_string(cell.quantity));
        } else {
            emptyCells.push_back(address);
        }
    }

    cout << "Общая загруженность: " 
         << fixed << setprecision(2) 
         << (totalItems / 6000.0 * 100) << "%\n";

    vector<char> zones = {'A', 'B', 'C'};
    for (char zone : zones) {
        double percent = (zoneItems[zone] / 2000.0) * 100;
        cout << "Зона " << zone << ": " << percent << "%\n";
    }

    cout << "\nЗаполненные ячейки (" << occupiedCells.size() << "):\n";
    for (const string& s : occupiedCells) cout << "  " << s << endl;

    cout << "\nПустые ячейки (" << emptyCells.size() << "):\n";
    for (const string& s : emptyCells) cout << "  " << s << endl;
}

int main() {
    initializeWarehouse();
    string line;

    cout << "=== Система управления складом ===\n"
         << "Доступные команды:\n"
         << "  ADD <товар> <кол-во> <адрес>\n"
         << "  REMOVE <товар> <кол-во> <адрес>\n"
         << "  INFO\n"
         << "Для выхода введите Ctrl+C\n\n";

    while (true) {
        cout << "> ";
        getline(cin, line);
        if (line.empty()) continue;

        istringstream iss(line);
        vector<string> tokens;
        string token;
        while (iss >> token) tokens.push_back(token);

        if (tokens.empty()) continue;

        string cmd = tokens[0];
        if (cmd == "ADD") processAddCommand(tokens);
        else if (cmd == "REMOVE") processRemoveCommand(tokens);
        else if (cmd == "INFO") processInfoCommand();
        else cout << "Ошибка: неизвестная команда \"" << cmd << "\"\n";
    }

    return 0;
}

File: laba5/ex1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Cell struct {
	item     string
	quantity int
}

var warehouse map[string]*Cell

func initializeWarehouse() {
	warehouse = make(map[string]*Cell)
	for _, zone := range []rune{'A', 'B', 'C'} {
		for shelf := 1; shelf <= 20; shelf++ {
			for section := 1; section <= 5; section++ {
				for rack := 1; rack <= 2; rack++ {
					address := fmt.Sprintf("%c%d%d%d", zone, shelf, section, rack)
					warehouse[address] = &Cell{"", 0}
				}
			}
		}
	}
}

func processAddCommand(tokens []string) {
	if len(tokens) != 4 {
		fmt.Println("Ошибка: Неверный формат. Используйте: ADD <товар> <кол-во> <адрес>")
		return
	}

	item := tokens[1]
	quantity, err := strconv.Atoi(tokens[2])
	if err != nil || quantity <= 0 {
		fmt.Println("Ошибка: Количество должно быть положительным числом")
		return
	}

	address := tokens[3]
	cell, exists := warehouse[address]
	if !exists {
		fmt.Printf("Ошибка: Ячейка %s не существует\n", address)
		return
	}

	if cell.quantity == 0 {
		if quantity > 10 {
			fmt.Println("Ошибка: В ячейку нельзя добавить более 10 единиц")
			return
		}
		cell.item = item
		cell.quantity = quantity
		fmt.Printf("Добавлено %d ед. товара \"%s\" в %s\n", quantity, item, address)
	} else {
		if cell.item != item {
			fmt.Printf("Ошибка: В ячейке находится другой товар (\"%s\")\n", cell.item)
			return
		}
		if cell.quantity+quantity > 10 {
			fmt.Printf("Ошибка: Превышен лимит. Доступно места: %d\n", 10-cell.quantity)
			return
		}
		cell.quantity += quantity
		fmt.Printf("Обновлено: %s x%d в %s\n", item, cell.quantity, address)
	}
}

func processRemoveCommand(tokens []string) {
	if len(tokens) != 4 {
		fmt.Println("Ошибка: Неверный формат. Используйте: REMOVE <товар> <кол-во> <адрес>")
		return
	}

	item := tokens[1]
	quantity, err := strconv.Atoi(tokens[2])
	if err != nil || quantity <= 0 {
		fmt.Println("Ошибка: Количество должно быть положительным числом")
		return
	}

	address := tokens[3]
	cell, exists := warehouse[address]
	if !exists {
		fmt.Printf("Ошибка: Ячейка %s не существует\n", address)
		return
	}

	if cell.quantity == 0 {
		fmt.Printf("Ошибка: Ячейка %s пуста\n", address)
		return
	}

	if cell.item != item {
		fmt.Printf("Ошибка: В ячейке находится товар \"%s\"\n", cell.item)
		return
	}

	if cell.quantity < quantity {
		fmt.Printf("Ошибка: Недостаточно товара. Доступно: %d\n", cell.quantity)
		return
	}

	cell.quantity -= quantity
	if cell.quantity == 0 {
		cell.item = ""
	}
	fmt.Printf("Удалено %d ед. товара \"%s\" из %s. Остаток: %d\n", quantity, item, address, cell.quantity)
}

func processInfoCommand() {
	total := 0
	zones := make(map[rune]int)
	var filled, empty []string

	for addr, cell := range warehouse {
		if cell.quantity > 0 {
			total += cell.quantity
			zones[rune(addr[0])] += cell.quantity
			filled = append(filled, fmt.Sprintf("%s: \"%s\" x%d", addr, cell.item, cell.quantity))
		} else {
			empty = append(empty, addr)
		}
	}

	fmt.Printf("Общая загруженность: %.2f%%\n", float64(total)/6000*100)
	for _, z := range []rune{'A', 'B', 'C'} {
		fmt.Printf("Зона %c: %.2f%%\n", z, float64(zones[z])/2000*100)
	}

	fmt.Printf("\nЗаполненные ячейки (%d):\n", len(filled))
	for _, f := range filled {
		fmt.Println(" ", f)
	}

	fmt.Printf("\nПустые ячейки (%d):\n", len(empty))
	for _, e := range empty {
		fmt.Println(" ", e)
	}
}

func main() {
	initializeWarehouse()
	fmt.Println("=== Система управления складом ===")
	fmt.Println("Формат команд:")
	fmt.Println("  ADD <товар> <кол-во> <адрес>")
	fmt.Println("  REMOVE <товар> <кол-во> <адрес>")
	fmt.Println("  INFO")

	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Print("> ")
		scanner.Scan()
		input := strings.Fields(scanner.Text())
		if len(input) == 0 {
			continue
		}

		switch input[0] {
		case "ADD":
			processAddCommand(input)
		case "REMOVE":
			processRemoveCommand(input)
		case "INFO":
			processInfoCommand()
		default:
			fmt.Println("Неизвестная команда")
		}
	}
}
